Client-3753: path expressions integration redux

sfalvo-aerospike opened the pull request:
Before opening the PR for the previous branch of this name, I asked about method renaming, and people seemed genuinely confused.  I didn't think anything of it; maybe I misheard.  Turns out I was right to ask -- the C code used obsolete naming conventions.  This (new/revived) branch aims to remedy this by switching names of operations and flags to match their equivalents within the PRD at https://aerospike.atlassian.net/wiki/spaces/CORE/pages/4347592705/PRD+-+Path+Expressions+Add+an+ability+to+select+multiple+elements+of+a+CDT+Map+List+using+expressions+for+context#%F0%9F%8F%8E%EF%B8%8F-Quickstart-Guide .

sfalvo-aerospike committed:
CLIENT-3753: Renaming methods to comply with PRD recommendations

sfalvo-aerospike committed:
CLIENT-3753: Another round of method renaming to comply with PRD recommendations.

juliannguyen4 commented on src/include/aerospike/as_cdt_ctx.h:
```diff
@@ -310,20 +310,30 @@ as_cdt_ctx_add_map_value(as_cdt_ctx* ctx, as_val* val)
  * Add expr to select ctx.  The ctx does NOT take ownership of exp.
+ * The passed expression must return a boolean.
+ *
+ * All children of the current level will be selected then the filter expression
```
```suggestion
 * All children of the current level will be selected, and then the filter expression
```

juliannguyen4 commented on src/include/aerospike/as_cdt_ctx.h:
```diff
@@ -310,20 +310,30 @@ as_cdt_ctx_add_map_value(as_cdt_ctx* ctx, as_val* val)
+ * The passed expression must return a boolean.
+ *
+ * All children of the current level will be selected then the filter expression
+ * applied to each item in turn.  Items that return true will be added to the
```
```suggestion
 * is applied to each item in turn.  Items that cause the expression to evaluate to true will be added to the
```

juliannguyen4 commented on src/include/aerospike/as_cdt_ctx.h:
```diff
@@ -310,20 +310,30 @@ as_cdt_ctx_add_map_value(as_cdt_ctx* ctx, as_val* val)
+ *
+ * All children of the current level will be selected then the filter expression
+ * applied to each item in turn.  Items that return true will be added to the
+ * list of items returned in a query for this level.  Items that return false
```
```suggestion
 * list of items returned in a query for this level.  Items that cause the expression to evaluate to false
```

juliannguyen4 commented on src/include/aerospike/as_exp.h:
```diff
@@ -2972,7 +2974,7 @@ as_exp_destroy_base64(char* base64)
  * @param Bin to which the application applies.
```
This is missing the parameter names; the other expression macros have them

For `__rtype`, it might help to include the enum name so downstream developers know what the valid values are:

` * @param __vtype		Value type (as_exp_type).`

sfalvo-aerospike replied:
I believe I've addressed most of this feedback in 3bf3928, if not everything.

sfalvo-aerospike replied:
I updated the tests to include negative testing for as_operations_select-/modify-by-path functions.  I'll try to add tests for as_exp_* as well next.

sfalvo-aerospike replied:
Ah, I cannot add negative tests for as_exp_select_by_path()/as_exp_modify_by_path(), because these macros directly build structures in memory.  There is no way to intercede if ctx==NULL.

juliannguyen4 commented on src/include/aerospike/as_operations.h:
```diff
@@ -793,18 +815,21 @@ typedef enum {
+ * Create CDT modification operation.
+ *
+ * The results of the evaluation of the modifying expression will replace the
+ * selected map and the changes written back to storage.
```
```suggestion
 * selected map, and the changes are written back to storage.
```

juliannguyen4 reviewed (changes requested):
left some suggestions.

Also, the PRD mentions that the operations `selectFromCdt` and `modifyCdt` must behave this way:

`It is an error to pass no items in the context list.`

It would be good to document and test this case

https://aerospike.atlassian.net/wiki/spaces/CORE/pages/4347592705/PRD+-+Path+Expressions+Add+an+ability+to+select+multiple+elements+of+a+CDT+Map+List+using+expressions+for+context#Step-0.1%3A-Stand-up-your-server-using…Docker-compose-%7C-Aerolab

sfalvo-aerospike committed:
CLIENT-3753: Use types to help document interfaces for flags.

In slack, there's some confusion over the validity of some AS_CDT_SELECT_xxx
flags when attempting to invoke a modify_cdt operation.  It turns out that the
AS_CDT_SELECT_xxx flags are (mostly!) irrelevant for modify_cdt operations.

Based on feedback from the Python client implementation, this patch introduces
a new enumeration for modify_cdt operation flags, as_cdt_modify_flags.  These
flags are intended to be equivalent to similar values in as_cdt_select_flags
where it makes sense, and to omit/redefine flags that are unique to modify_cdt
operations as appropriate.

It is believed that this will help authors (particularly of language bindings)
disambiguate how the flags field is intended to be used between various
select_from_cdt and modify_cdt functions.

sfalvo-aerospike committed:
CLIENT-3753: Alter flag parameters to use enums instead of uint32_t.

This change aids in procedure documentation, and might even aid in compiler
automated error detection.

sfalvo-aerospike committed:
Update src/include/aerospike/as_cdt_ctx.h

Co-authored-by: Julian Nguyen <109386615+juliannguyen4@users.noreply.github.com>

sfalvo-aerospike committed:
Update src/include/aerospike/as_cdt_ctx.h

Co-authored-by: Julian Nguyen <109386615+juliannguyen4@users.noreply.github.com>

sfalvo-aerospike committed:
Update src/include/aerospike/as_cdt_ctx.h

Co-authored-by: Julian Nguyen <109386615+juliannguyen4@users.noreply.github.com>

sfalvo-aerospike committed:
Update src/include/aerospike/as_operations.h

Co-authored-by: Julian Nguyen <109386615+juliannguyen4@users.noreply.github.com>

sfalvo-aerospike committed:
CLIENT-3753: doc string updates for consistency

fallguy reviewed (changes requested):
The method names do not seem to be correct. You can see the correct method names here: https://github.com/aerospike-examples/path-expressions-java-preview.

sfalvo-aerospike committed:
CLIENT-3753: Method rename, redux redux

Previous rename patches assumed the PRD was up to date.  Brian P reported that
the PRD was obsolete and to use
https://github.com/aerospike-examples/path-expressions-java-preview
as the reference document instead.

juliannguyen4 commented on src/include/aerospike/as_exp.h:
```diff
@@ -2967,11 +2969,13 @@ as_exp_destroy_base64(char* base64)
+ * @param __rtype    Return type specifier (e.g., AS_EXP_TYPE_MAP).
+ * @param __mod_exp  Expression to apply.
+ * @param __flags    Flags (see enum as_cdt_modify_flags).
+ * @param __bin      Bin to which the application applies.
```
```suggestion
 * @param __bin      Bin expression to which the application applies.
```

Forgot to mention -- I believe the bin parameter should be a bin expression; if I try to pass in a string, I think I get error code 4 from the server

If someone else can confirm, that would be great

sfalvo-aerospike replied:
I believe you're right, because it also sets the type of the bin as well.  Thanks!

sfalvo-aerospike committed:
CLIENT-3753: Rename "built-in vars" to "loop vars"

Per Preview document.

juliannguyen4 commented on src/include/aerospike/as_operations.h:
```diff
@@ -827,29 +827,37 @@ typedef enum {
 } as_cdt_modify_flags;
```
Looked at this again, and I think since the only option is to "fail" or "not fail", we can have another enum value called `AS_CDT_MODIFY_FAIL = 0x0` so customers know how to make this operation fail in this case?

sfalvo-aerospike replied:
Would this set a precedent where any future flag AS_CDT_MODIFY_XYZ = 0x20 (say) has a corresponding antonym AS_CDT_MODIFY_NOTXYZ = 0?  I mention this because @fallguy and I had a quick huddle today where I've been asked to flesh out some of the Quick Start coding examples as documentation.  I think these docs would be sufficient to disambiguate a fail vs. no-fail behavior.  @fallguy do you have any input on this matter?

juliannguyen4 replied:
Alternatively we could document setting the flags parameter to `0` to enable the default behavior

sfalvo-aerospike replied:
Added a AS_CDT_MODIFY_DEFAULT (= 0) flag and documented it.  What do you think?

sfalvo-aerospike committed:
Update src/include/aerospike/as_exp.h

Co-authored-by: Julian Nguyen <109386615+juliannguyen4@users.noreply.github.com>

sfalvo-aerospike committed:
Update as_exp.h

sfalvo-aerospike committed:
CLIENT-3753: Add tests for empty and NULL contexts

sfalvo-aerospike committed:
CLIENT-3753: Re-org code to compile

sfalvo-aerospike committed:
CLIENT-3753: Add a default behavior option.

sfalvo-aerospike committed:
CLIENT-3753: Refine docs per feedback

mirzakaracic reviewed (commented):
Looks good

sfalvo-aerospike committed:
CLIENT-3753: Fix double-free bug.

sfalvo-aerospike committed:
CLIENT-3753: Consider Shu's feedback for symbol naming

sfalvo-aerospike committed:
CLIENT-3841: Introduce bool and blob loopvar types

sfalvo-aerospike committed:
CLIENT-3841: Change "builtin" naming to "loopvar" naming

sfalvo-aerospike committed:
CLIENT-3841: Missed a symbol to rename

sfalvo-aerospike commented:
Rebased branch to unblock Julian

sfalvo-aerospike committed:
CLIENT-3753: Make private flag private in API docs as well.

sfalvo-aerospike committed:
CLIENT-3753: Document apt command to install deps for Debian make package

sfalvo-aerospike committed:
CLIENT-3753: Final symbol renaming

This patch also includes WIP for as_exp_result_remove tests.  In their current
state, these tests should not negatively affect test runs, since intended
outcomes haven't been asserted yet.  This test will be completed in a future
patch.

sfalvo-aerospike committed:
CLIENT-3753: Fix wrong path expression result type

sfalvo-aerospike committed:
CLIENT-3753: Remove result_remove test entirely.

I need to start fresh with this.  This test is just not working right for me.

juliannguyen4 commented on src/test/aerospike_list/list_basics.c:
```diff
@@ -3273,14 +3270,31 @@ TEST(list_select, "test select")
-	as_operations ops;
+	assert_false(as_operations_select_by_path(&ops, BIN_NAME, &ctx, 0));
+
+	// Back to positive testing
```
Possible to split negative and positive tests into their own test cases?

sfalvo-aerospike replied:
Per previous discussion, yes; but it will be in a separate PR.

juliannguyen4 commented on src/test/aerospike_list/list_basics.c:
Maybe rename this test case so we know that it's testing iteration on list values (seems like that's the main goal for this test)

sfalvo-aerospike replied:
I have reviewed the code, and list_select makes sense to me (since it's using CDT select for the query).  What name would make most sense for you?

juliannguyen4 replied:
Looking back, the test file is already named `list_basics.c`, so I guess it's fine to leave the naming as-is. This is just a nit

juliannguyen4 commented on src/test/aerospike_list/list_basics.c:
```diff
@@ -3273,14 +3270,31 @@ TEST(list_select, "test select")
-	as_operations_cdt_select(&ops, BIN_NAME, &ctx, 0);
+
+	as_cdt_ctx_inita(&ctx, 2);
+	as_cdt_ctx_add_all_children_with_filter(&ctx, exp1);
```
Since exp1 always returns `true`, we can remove this filter entirely, right?

sfalvo-aerospike replied:
No, it is still required.  exp1 is used for selecting the entirety of the first tier of results, while exp2 is used for selecting only portions of the children of the 2nd tier.  I know it's confusing, but it would make more sense if I could draw a tree diagram and illustrate where exp1 and exp2 applies.

juliannguyen4 commented on src/test/aerospike_list/list_basics.c:
Maybe rename "book" to "books" so it's more clear that the map entry maps to a list of books, not a single book

sfalvo-aerospike replied:
This should be fixed in the next PR.  I've replaced all the hard-coded "book" strings with a single macro BOOKS.

juliannguyen4 commented on src/test/aerospike_list/list_basics.c:
```diff
@@ -3371,24 +3385,27 @@ TEST(list_select2, "test select")
-	as_cdt_ctx_add_exp(&ctx, exp1);
-	as_cdt_ctx_add_exp(&ctx, exp2);
-	as_cdt_ctx_add_exp(&ctx, exp3);
+	as_cdt_ctx_add_all_children_with_filter(&ctx, exp1);
```
See comment https://github.com/aerospike/aerospike-client-c/pull/177#r2519138854

sfalvo-aerospike replied:
I'm not sure what comment you're linking to here.  If I click on the link, GitHub just ignores the click.  Can you quote the concern as re response inline here?  Thanks!

juliannguyen4 commented on src/test/aerospike_list/list_basics.c:
```diff
@@ -3371,24 +3385,27 @@ TEST(list_select2, "test select")
-		stack_check_list_elt_0 = alloca(strlen(as_string_get(heap_check_list_elt_0)) + 1);
-		strcpy(stack_check_list_elt_0, as_string_get(heap_check_list_elt_0));
-		as_string_destroy(heap_check_list_elt_0);
+		if (check_list_size > 0) {
```
Shouldn't we assert that check_list_size is 2?

sfalvo-aerospike replied:
Not here, no.  The purpose of this code is to prevent a false positive in Valgrind when the test fails.  If you look below, around line 3422, I do explicitly check to see that the list length is 2.  But for this code, I don't care about the size of the actual list; I just want to copy the list as it already is.

juliannguyen4 commented on src/test/aerospike_list/list_basics.c:
Might be better to check the whole string (book name) using strcmp, and to check both book names instead of one of them

juliannguyen4 commented on src/test/aerospike_list/list_basics.c:
Should be updated to say "modify" instead of "apply"

sfalvo-aerospike replied:
Done

juliannguyen4 reviewed (approved):
Left some suggestions about naming and test organization. Otherwise LGTM

juliannguyen4 commented:
@sfalvo-aerospike I believe we're also missing documentation for this struct in the C client:
```C
typedef enum {
	AS_EXP_LOOPVAR_KEY = 0,
	AS_EXP_LOOPVAR_VALUE = 1,
	AS_EXP_LOOPVAR_INDEX = 2
} as_exp_loopvar_type;
```

juliannguyen4 commented on src/include/aerospike/as_exp.h:
```diff
@@ -2965,16 +2965,19 @@ as_exp_destroy_base64(char* base64)
+ *                 the context be empty.
+ * @param __rtype  Return type specifier (e.g., AS_EXP_TYPE_MAP).
+ * @param __flags  Flags (see enum as_exp_path_select_flags).
+ * @param __bin    Bin expression to which the query applies.
```
```suggestion
 * @param __bin    Bin expression to which this expression applies.
```

juliannguyen4 commented on src/include/aerospike/as_exp.h:
```diff
@@ -2984,17 +2987,22 @@ as_exp_destroy_base64(char* base64)
 /**
- * Constructs a CDT apply operation.  This can be used to perform in-place
- * updates to bins across a number of records or fields of records.
+ * Constructs a CDT apply operation.
```
```suggestion
 * Constructs an apply by path operation.
```

juliannguyen4 commented on src/include/aerospike/as_exp.h:
```diff
@@ -2984,17 +2987,22 @@ as_exp_destroy_base64(char* base64)
+ * @param __rtype    Return type specifier (e.g., AS_EXP_TYPE_MAP).
+ * @param __mod_exp  Expression to apply.
+ * @param __flags    Flags (see enum as_exp_path_modify_flags).
+ * @param __bin      Bin expression to which the application applies.
```
```suggestion
 * @param __bin      Bin expression to which this expression applies.
```

sfalvo-aerospike replied:
There are several expressions provided in the query; I need to be specific to which expression is affected.

juliannguyen4 reviewed (commented):
More comments

juliannguyen4 commented on src/include/aerospike/as_operations.h:
```diff
@@ -776,35 +776,120 @@ as_operations_add_touch(as_operations* ops);
+	/**
+	 * If the expression in the context hits an invalid type (e.g., selects
+	 * as an integer when the value is a string), do not fail the operation;
+	 * just ignore those elements.  Interpret UNKNOWN as false instead.
```
```suggestion
	 * just ignore those elements.  Interpret an expression that returns UNKNOWN as false instead.
```

Last sentence was a bit confusing for me.

juliannguyen4 commented on src/include/aerospike/as_operations.h:
```diff
@@ -776,35 +776,120 @@ as_operations_add_touch(as_operations* ops);
+	AS_EXP_PATH_SELECT_MAP_KEY = 2,
+
+	/**
+	 * Returns the list of map (key, value) pairs of the nodes finally selected
```
When testing this, EXP_PATH_SELECT_MAP_KEY_VALUE returns a list of "flattened" key-value pairs, meaning the list is formatted like `[key1, value1, key2, value2, ...]` where the map looks like `{key1: value1, key2: value2, ...}`. It's not clear in the documentation that this is the expected behavior, so we should mention this

juliannguyen4 commented on src/include/aerospike/as_exp.h:
```diff
@@ -2965,16 +2965,19 @@ as_exp_destroy_base64(char* base64)
- * @param Bin to which the query applies.
+ * @param __ctx    Pointer to a CDT context.  This cannot be NULL, nor can
+ *                 the context be empty.
+ * @param __rtype  Return type specifier (e.g., AS_EXP_TYPE_MAP).
```
@ajkimdev pointed out that  this parameter should be named `__vtype` ("value type") to be consistent with the other expressions that have a parameter that takes in the return value type of the expression

sfalvo-aerospike committed:
Merge branch 'stage' into client-3753-path-expressions-integration

sfalvo-aerospike committed:
CLIENT-3753: Apply github feedback

sfalvo-aerospike committed:
CLIENT-3753: Apply feedback

sfalvo-aerospike commented:
> @sfalvo-aerospike I believe we're also missing documentation for this struct in the C client:
> 
> ```c
> typedef enum {
> 	AS_EXP_LOOPVAR_KEY = 0,
> 	AS_EXP_LOOPVAR_VALUE = 1,
> 	AS_EXP_LOOPVAR_INDEX = 2
> } as_exp_loopvar_type;
> ```

There are a large number of other enums which are also undocumented as well, all of which appear to be serialization encodings.  Considering the proximity of these enums, it appears that these are also intended to be serialization constants, and not specifically intended to be used directly by the API consumer.

BrianNichols reviewed (changes requested):
In as_operations.h, the new enums need to be moved to the "Types" section.

In exp_operate.c, put the new extern declaration outside of the function and into the "GLOBAL VARS" section.

In README.md, merge the new "For Debian Linux" lines into the already existing "Debian and Ubuntu" in the same file.

sfalvo-aerospike committed:
CLIENT-3753: Applying review feedback

BrianNichols committed:
Fix double to float conversion warnings.
